### PR TITLE
Improve long press menu

### PR DESCRIPTION
**Pull request by mueller-ma**
* Show a menu with the possible actions: write item update tag, write
sitemap tag or create shortcut
* Show "custom" state for item update tags

Signed-off-by: mueller-ma <mueller-ma@users.noreply.github.com>

### CODE REVIEW

**Comment by maniac103** [comment]
I really don't like two dialogs opening subsequently. I think instead of the dialogs we should use (nested) context menus. RecyclerView doesn't support those out of the box, but I have coded this in the past: https://github.com/slapperwan/gh4a/blob/master/app/src/main/java/com/gh4a/widget/ContextMenuAwareRecyclerView.java ... the sub menu will probably be a bit tricky (see e.g. http://www.blackmoonit.com/2010/09/using-submenu-in-listactivitys-context/) but I think it will provide much better UX.

**Comment by mueller-ma** [comment]
I downloaded OctoDroid. Can you tell me, where this dialogs are used?

**Comment by maniac103** [comment]
E.g. in the news feed (default start view if you're logged in) ... on long press, options for visiting the respective repo, user or issue are offered.

**Comment by mueller-ma** [comment]
This menu doesn't seem to have any submenus. What about the menu when clicking on the three dots right to a comment and than "Add reaction"?

**Comment by maniac103** [comment]
That's what I meant by 'tricky' ... that part doesn't have a ready-to-use example yet. Inflation of a sub menu should work the same way in context menus and popup menus though. The code handling the menu you mentioned can be found here: https://github.com/slapperwan/gh4a/blob/master/app/src/main/java/com/gh4a/adapter/CommitNoteAdapter.java#L272-L277

**Comment by mueller-ma** [comment]
> but I think it will provide much better UX.

IMO the UX might be a bit better, but I don't see why two menus are bad UI.

**Comment by maniac103** [comment]
No one said it is. It's fine in that it isn't confusing and is usable, but it just feels unpolished and clunky to mme

**Comment by maniac103** [comment]
PoC implementation of context menu instead of AlertDialogs: https://github.com/openhab/openhab-android/commit/7e8f9216dd40623e65fcc2837eb92065c6801abd

**Comment by mueller-ma** [comment]
I agree that it's looking more polished, but the context menu is much smaller. The title of the NFC state selection menu is cut and most of the time that state menu must be scrolled to view all possible states.
Removing the two NFC actions if NFC isn't available is better UX, but it makes it impossible for UI tests to check for this menu. Emulators don't have NFC support. Maybe we can add some debug switch.

**Comment by maniac103** [comment]
> I agree that it's looking more polished, but the context menu is much smaller. The title of the NFC state selection menu is cut and most of the time that state menu must be scrolled to view all possible states.

I don't see the scrolling as a big issue. The title in the NFC menu can be shortened, as it's clear what the menu is about due to the first level item.

> Removing the two NFC actions if NFC isn't available is better UX, but it makes it impossible for UI tests to check for this menu. Emulators don't have NFC support. Maybe we can add some debug switch.

Yeah, we can force it by checking for emulator, or maybe we can leverage https://developer.android.com/reference/kotlin/android/app/ActivityManager?hl=en#isuseramonkey ... if neither works nicely, we can still use a resource boolean or test-only code overlay or something. All of this shouldn't be a blocker.

**Comment by mueller-ma** [comment]
> The title in the NFC menu can be shortened, as it's clear what the menu is about due to the first level item.

The first level isn't present if the widget doesn't have a linked page.

**Comment by maniac103** [comment]
> The first level isn't present if the widget doesn't have a linked page. And even if we shorten the English string, translations might be longer.

Then we also can shorten translations. TBH, the menu title always felt *way* too long and verbose for me, especially given there's an activity triggered by the menu items which doesn't do anything by itself.
We could make the title e.g. 'Write to NFC tag' and show a text in the NFC activity among the lines of 'Touch NFC tag to make it send the command XY to Item YZ' or 'Touch NFC tag to make it open the app with page XY'.

**Comment by mueller-ma** [comment]
I shortened the labels, should be RTM now.

File: mobile/src/main/java/org/openhab/habdroid/ui/WidgetListFragment.kt [hunk 1]
```diff
@@ -140,41 +154,118 @@ class WidgetListFragment : Fragment(), WidgetAdapter.ItemClickListener {
...
-                    } else if (linkToHandle != null) {
-                        startActivity(WriteTagActivity.createSitemapNavigationIntent(context, linkToHandle))
+    private fun populateNfcStatesMenu(
+        menu: Menu,
+        context: Context,
+        suggestedCommands: SuggestedCommandsFactory.SuggestedCommands,
+        widget: Widget
+    ) {
+        val name = widget.item?.name ?: return
+        val listener = object: MenuItem.OnMenuItemClickListener {
```
Comment (mueller-ma) [inline mobile/src/main/java/org/openhab/habdroid/ui/WidgetListFragment.kt]: False positive?

Comment (maniac103) [reply mobile/src/main/java/org/openhab/habdroid/ui/WidgetListFragment.kt]: You mean the Stickler error? Doesn't look like a false positive to me, as there's a rule to put a space before and after the colon in the `DerivedClassName : BaseClassName` syntax.

File: mobile/src/main/java/org/openhab/habdroid/ui/WidgetListFragment.kt [hunk 2]
```diff
@@ -140,41 +154,118 @@ class WidgetListFragment : Fragment(), WidgetAdapter.ItemClickListener {
...
+    private fun isEmulator(): Boolean {
+        return Build.FINGERPRINT.startsWith("generic") ||
+            Build.FINGERPRINT.startsWith("unknown") ||
+            Build.MODEL.contains("google_sdk") ||
+            Build.MODEL.contains("Emulator") ||
+            Build.MODEL.contains("Android SDK built for x86") ||
+            Build.MANUFACTURER.contains("Genymotion") ||
+            (Build.BRAND.startsWith("generic") && Build.DEVICE.startsWith("generic")) ||
+            "google_sdk" == Build.PRODUCT
+    }
```
Comment (maniac103) [inline mobile/src/main/java/org/openhab/habdroid/ui/WidgetListFragment.kt]: Maybe put this to `Util`?

File: mobile/src/main/java/org/openhab/habdroid/ui/WidgetListFragment.kt [hunk 3]
```diff
@@ -140,41 +154,118 @@ class WidgetListFragment : Fragment(), WidgetAdapter.ItemClickListener {
...
+                val id = item?.itemId ?: return false
+                if (id == CONTEXT_MENU_ID_WRITE_CUSTOM_TAG) {
+                    val input = EditText(context)
+                    input.inputType = suggestedCommands.inputTypeFlags
+                    val customDialog = AlertDialog.Builder(context)
+                        .setTitle(getString(R.string.item_picker_custom))
+                        .setView(input)
+                        .setPositiveButton(android.R.string.ok) { _, _ ->
+                            startActivity(WriteTagActivity.createItemUpdateIntent(context, name, input.text.toString(),
+                                input.text.toString(), widget.item.label.orEmpty())) }
```
Comment (maniac103) [inline mobile/src/main/java/org/openhab/habdroid/ui/WidgetListFragment.kt]: Shouldn't the closing `}` be on a new line?

**Comment by mueller-ma** [comment]
The tests pass on my local machine with an api 22 emulator.

**Pull request merged**